touch/touch.1
.Dd December 2008
.Dt TOUCH 1

.Sh NAME

.Nm touch
.Nd change file access and modification times

.Sh SYNOPSIS

.Nm touch
.Ar [-acm][ -r ref_file| -t time] file...

.Sh DESCRIPTION

    The touch utility shall change the modification times, access times, or
both of files. The modification time shall be equivalent to the value of the
st_mtime member of the stat structure for a file, as described in the System
Interfaces volume of IEEE Std 1003.1-2001; the access time shall be
equivalent to the value of st_atime.

    The time used can be specified by the -t time option-argument, the
corresponding time fields of the file referenced by the -r ref_file
option-argument, or the date_time operand, as specified in the following
sections. If none of these are specified, touch shall use the current time
(the value returned by the equivalent of the time() function defined in the
System Interfaces volume of IEEE Std 1003.1-2001).

    For each file operand, touch shall perform actions equivalent to the
following functions defined in the System Interfaces volume of IEEE Std
1003.1-2001:

        If file does not exist, a creat() function call is made with the file
operand used as the path argument and the value of the bitwise-inclusive OR
of S_IRUSR, S_IWUSR, S_IRGRP, S_IWGRP, S_IROTH, and S_IWOTH used as the mode
argument.

        The utime() function is called with the following arguments:

            The file operand is used as the path argument.

            The utimbuf structure members actime and modtime are determined
as described in the OPTIONS section.

.Sh OPTIONS

    The touch utility shall conform to the Base Definitions volume of IEEE
Std 1003.1-2001, Section 12.2, Utility Syntax Guidelines.

    The following options shall be supported:

    -a
        Change the access time of file. Do not change the modification time
unless -m is also specified.
    -c
        Do not create a specified file if it does not exist. Do not write any
diagnostic messages concerning this condition.
    -m
        Change the modification time of file. Do not change the access time
unless -a is also specified.
    -r  ref_file
        Use the corresponding time of the file named by the pathname ref_file
instead of the current time.
    -t  time
        Use the specified time instead of the current time. The
option-argument shall be a decimal number of the form:

        [[CC]YY]MMDDhhmm[.SS]

        where each two digits represents the following:

        MM
            The month of the year [01,12].
        DD
            The day of the month [01,31].
        hh
            The hour of the day [00,23].
        mm
            The minute of the hour [00,59].
        CC
            The first two digits of the year (the century).
        YY
            The second two digits of the year.
        SS
            The second of the minute [00,60].

        Both CC and YY shall be optional. If neither is given, the current
year shall be assumed. If YY is specified, but CC is not, CC shall be derived
as follows:

        If YY is:
        	

        CC becomes:

        [69,99]
        	

        19

        [00,68]
        	

        20

        Note:
            It is expected that in a future version of IEEE Std 1003.1-2001
the default century inferred from a 2-digit year will change. (This would
apply to all commands accepting a 2-digit year as input.)

        The resulting time shall be affected by the value of the TZ
environment variable. If the resulting time value precedes the Epoch, touch
shall exit immediately with an error status. The range of valid times past
the Epoch is implementation-defined, but it shall extend to at least the time
0 hours, 0 minutes, 0 seconds, January 1, 2038, Coordinated Universal Time.
Some implementations may not be able to represent dates beyond January 18,
2038, because they use signed int as a time holder.

        The range for SS is [00,60] rather than [00,59] because of leap
seconds. If SS is 60, and the resulting time, as affected by the TZ
environment variable, does not refer to a leap second, the resulting time
shall be one second after a time where SS is 59. If SS is not given a value,
it is assumed to be zero.

    If neither the -a nor -m options were specified, touch shall behave as if
both the -a and -m options were specified.

.Sh OPERANDS

    The following operands shall be supported:

    file
        A pathname of a file whose times shall be modified.

.Sh STDIN

    Not used.

.Sh INPUT FILES

    None.

.Sh ENVIRONMENT VARIABLES

    The following environment variables shall affect the execution of touch:

    LANG
        Provide a default value for the internationalization variables that
are unset or null. (See the Base Definitions volume of IEEE Std 1003.1-2001,
Section 8.2, Internationalization Variables for the precedence of
internationalization variables used to determine the values of locale
categories.)
    LC_ALL
        If set to a non-empty string value, override the values of all the
other internationalization variables.
    LC_CTYPE
        Determine the locale for the interpretation of sequences of bytes of
text data as characters (for example, single-byte as opposed to multi-byte
characters in arguments).
    LC_MESSAGES
        Determine the locale that should be used to affect the format and
contents of diagnostic messages written to standard error.
    NLSPATH
        [XSI] [Option Start] Determine the location of message catalogs for
the processing of LC_MESSAGES . [Option End]
    TZ
        Determine the timezone to be used for interpreting the time
option-argument. If TZ is unset or null, an unspecified default timezone
shall be used.

.Sh ASYNCHRONOUS EVENTS

    Default.

.Sh STDOUT

    Not used.

.Sh STDERR

    The standard error shall be used only for diagnostic messages.

.Sh OUTPUT FILES

    None.

.Sh EXTENDED DESCRIPTION

    None.

.Sh EXIT STATUS

    The following exit values shall be returned:

     0
        The utility executed successfully and all requested changes were
made.
    >0
        An error occurred.

.Sh CONSEQUENCES OF ERRORS

    Default.

The following sections are informative.
.Sh APPLICATION USAGE

    The interpretation of time is taken to be seconds since the Epoch (see
the Base Definitions volume of IEEE Std 1003.1-2001, Section 4.14, Seconds
Since the Epoch). It should be noted that implementations conforming to the
System Interfaces volume of IEEE Std 1003.1-2001 do not take leap seconds
into account when computing seconds since the Epoch. When SS=60 is used, the
resulting time always refers to 1 plus seconds since the Epoch for a time
when SS=59.

    Although the -t time option-argument specifies values in 1969, the access
time and modification time fields are defined in terms of seconds since the
Epoch (00:00:00 on 1 January 1970 UTC). Therefore, depending on the value of
TZ when touch is run, there is never more than a few valid hours in 1969 and
there need not be any valid times in 1969.

    One ambiguous situation occurs if -t time is not specified, -r ref_file
is not specified, and the first operand is an eight or ten-digit decimal
number. A portable script can avoid this problem by using:

    touch -- file

    or:

    touch ./file

    in this case.

.Sh EXAMPLES

    None.

.Sh RATIONALE

    The functionality of touch is described almost entirely through
references to functions in the System Interfaces volume of IEEE Std
1003.1-2001. In this way, there is no duplication of effort required for
describing such side effects as the relationship of user IDs to the user
database, permissions, and so on.

    There are some significant differences between the touch utility in this
volume of IEEE Std 1003.1-2001 and those in System V and BSD systems. They
are upwards-compatible for historical applications from both implementations:

        In System V, an ambiguity exists when a pathname that is a decimal
number leads the operands; it is treated as a time value. In BSD, no time
value is allowed; files may only be touched to the current time. The -t time
construct solves these problems for future conforming applications (note that
the -t option is not historical practice).

        The inclusion of the century digits, CC, is also new. Note that a
ten-digit time value is treated as if YY, and not CC, were specified. The
caveat about the range of dates following the Epoch was included as
recognition that some implementations are not able to represent dates beyond
18 January 2038 because they use signed int as a time holder.

    The -r option was added because several comments requested this
capability. This option was named -f in an early proposal, but was changed
because the -f option is used in the BSD version of touch with a different
meaning.

    At least one historical implementation of touch incremented the exit code
if -c was specified and the file did not exist. This volume of IEEE Std
1003.1-2001 requires exit status zero if no errors occur.

.Sh FUTURE DIRECTIONS

    Applications should use the -r or -t options.

.Sh SEE ALSO

    date, the System Interfaces volume of IEEE Std 1003.1-2001, creat(),
time(), utime(), the Base Definitions volume of IEEE Std 1003.1-2001,
<sys/stat.h>



NEWS

Version 5 - unreleased

	- New utilities: cmp, tput, tabs, stty, uname, uniq

	- Bug fixes.

Version 4 - December 9, 2006

	- import codebase into git repository on git.kernel.org
	- fix compiler warnings on newer compilers
	- add x86-64 support
	- add WIP utility: pax
	- add utilities: env, id, strings, time
	- add simple "it works" tests
	- create per-util status files
	- drop hand-rolled 'compress', import BSD 'compress' instead

Version 3 - February 7, 2005

	- Add copyright/license info to each source file.
	- Make sure bug report address points to SourceForge.

	- New utilities: grep, pathchk, pwd, split, tty.
	- New work-in-progress (WIP) utility: mv.

	- Fix many 'compress' bugs.
	- Security fixes.

	- (internal) Move compress engine into lib.

Version 2 - January 26, 2005

	- Update README with project goals/manifesto,
	  and some URLs.

	- New utilities: compress, comm, date, what, wc, renice,
	  ipcrm, kill, rm.

	- Bug fixes.

Version 1 - August 21, 2004

	- Initial release.



touch/touch.cc
/*
 * Copyright 2004-2006 Jeff Garzik <jgarzik@pobox.com>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; see the file COPYING.  If not, write to
 * the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 */


#ifndef HAVE_CONFIG_H
#error missing autoconf-generated config.h.
#endif
#include "posixutils-config.h"

#include <sys/types.h>
#include <sys/stat.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <errno.h>
#include <time.h>
#include <utime.h>
#include <fcntl.h>
#include <libpu.h>

#define PFX "touch: "

static const char doc[] =
N_("touch - change file timestamps");

#define args_doc file_args_doc

static struct argp_option options[] = {
	{ NULL, 'a', NULL, 0,
	  N_("Change the access time of file") },
	{ "no-create", 'c', NULL, 0,
	  N_("Do not create a specified file if it does not exist") },
	{ NULL, 'm', NULL, 0,
	  N_("Change the modification time of file") },
	{ "reference", 'r', "FILE", 0,
	  N_("Use the corresponding time of FILE instead of the current time") },
	{ NULL, 't', "STAMP", 0,
	  N_("Use the specified time STAMP instead of the current time") },
	{ }
};

static int touch_init(struct walker *w, int argc, char **argv);
static int touch_pre_walk(struct walker *w);
static int touch_fn_actor(struct walker *w, const char *fn, const struct stat *lst);
static error_t parse_opt (int key, char *arg, struct argp_state *state);
static const struct argp argp = { options, parse_opt, args_doc, doc };

static struct walker walker;

static bool opt_default = true;
static bool opt_atime;
static bool opt_creat_ok = true;
static bool opt_mtime;
static bool opt_ref_file;
static time_t touch_atime, touch_mtime, touch_time;
static struct stat touch_st;


static int touch_fn_actor(struct walker *w, const char *fn, const struct stat *lst)
{
	struct utimbuf ut;
	struct stat st;
	int fd, second_time = 0;

	ut.actime = touch_atime;
	ut.modtime = touch_mtime;

	if (!opt_atime || !opt_mtime) {
		if (stat(fn, &st) < 0)
			goto err_out;

		if (!opt_atime)
			ut.actime = st.st_atime;
		if (!opt_mtime)
			ut.modtime = st.st_mtime;
	}

again_butthead:
	if (utime(fn, &ut) == 0)
		return 0;
	if (errno != ENOENT)
		goto err_out;
	if (!opt_creat_ok || second_time)
		return 1;

	fd = creat(fn, 0666);
	if (fd < 0)
		goto err_out;

	second_time = 1;
	goto again_butthead;

err_out:
	perror(fn);
	return 1;
}

static void parse_user_time(char *ut)
{
	struct tm tm;
	char *suff;
	int rc = 0, req_tokens = 99;

	memset(&tm, 0, sizeof(tm));

	suff = (char *) strchr(ut, '.');
	if (suff) {
		*suff = 0;
		suff++;
	}

	touch_time = time(NULL);
	if (strlen(ut) == 8) {
		localtime_r(&touch_time, &tm);
		req_tokens = 4;
		rc = sscanf(ut, "%02d%02d%02d%02d",
			    &tm.tm_mon,
			    &tm.tm_mday,
			    &tm.tm_hour,
			    &tm.tm_min);
	}
	else if (strlen(ut) == 10) {
		req_tokens = 5;
		rc = sscanf(ut, "%02d%02d%02d%02d%02d",
			    &tm.tm_year,
			    &tm.tm_mon,
			    &tm.tm_mday,
			    &tm.tm_hour,
			    &tm.tm_min);
		if (rc == 5) {
			if (tm.tm_year >= 69)
				tm.tm_year += 1900;
			else
				tm.tm_year += 2000;
		}
	}
	else if (strlen(ut) == 12) {
		req_tokens = 5;
		rc = sscanf(ut, "%04d%02d%02d%02d%02d",
			    &tm.tm_year,
			    &tm.tm_mon,
			    &tm.tm_mday,
			    &tm.tm_hour,
			    &tm.tm_min);
	}

	if (rc != req_tokens) {
		fprintf(stderr, PFX "invalid time spec\n");
		exit(1);
	}

	if (suff) {
		req_tokens = 1;
		rc = sscanf(suff, "%02d", &tm.tm_sec);
	}

	if (rc != req_tokens) {
		fprintf(stderr, PFX "invalid time spec\n");
		exit(1);
	}

	touch_time = mktime(&tm);
}

static error_t parse_opt (int key, char *arg, struct argp_state *state)
{
	PU_OPT_BEGIN

	case 'a':
		opt_default = false;
		opt_atime = true;
		break;
	case 'c':
		opt_creat_ok = false;
		break;
	case 'm':
		opt_default = false;
		opt_mtime = true;
		break;
	case 'r':
		if (stat(arg, &touch_st) < 0) {
			perror(arg);
			exit(1);
		}
		opt_ref_file = true;
		break;
	case 't':
		opt_ref_file = false;
		parse_user_time(optarg);
		break;

	PU_OPT_ARG
	PU_OPT_DEFAULT
	PU_OPT_END
}

static int touch_init(struct walker *w, int argc, char **argv)
{
	pu_init();
	touch_time = time(NULL);
	return 0;
}

static int touch_pre_walk(struct walker *w)
{
	if (opt_default)
		opt_mtime = opt_atime = true;

	if (opt_ref_file) {
		touch_atime = touch_st.st_atime;
		touch_mtime = touch_st.st_mtime;
	} else {
		touch_atime = touch_time;
		touch_mtime = touch_time;
	}

	return 0;
}

int main (int argc, char *argv[])
{
	walker.argp			= &argp;
	walker.init			= touch_init;
	walker.pre_walk			= touch_pre_walk;
	walker.cmdline_arg		= touch_fn_actor;
	return walk(&walker, argc, argv);
}
